core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -127,7 +128,14 @@ private[deploy] class Master(
     logInfo(s"Running Spark version ${org.apache.spark.SPARK_VERSION}")
     webUi = new MasterWebUI(this, webUiPort)
     webUi.bind()
-    masterWebUiUrl = "http://" + masterPublicAddress + ":" + webUi.boundPort
+    if (reverseProxy) {
+      masterWebUiUrl = conf.get("spark.ui.reverseProxyUrl", null)
```

Comment:
It is used in case when you are running spark master itself behind a proxy e.g. Oauth2 to provide authentication/authorization. Its to make sure "Back to Master"  link works when you are on workers UI.